design/one-pager-package-runtime-config.md

```diff
@@ -0,0 +1,222 @@
+# Package Runtime Config
+
+* Owner: Nic Cope (@negz)
+* Reviewers: Hasan Turken (@turkenh), Dan Mangum (@hasheddan)
+* Status: Accepted
+
+## Background
+
+Crossplane Providers are Kubernetes controller managers. They're packaged in an
+OCI container and expect to connect to an API server. You install a Provider
+declaratively by creating a Provider resource. When you do, the Crossplane
+package manager installs your provider by creating a Kubernetes Deployment,
+along with some accoutrements like a ServiceAccount and a Service. The RBAC
+manager also creates some RBAC ClusterRoles and ClusterRoleBindings to allow the
+Provider's ServiceAccount (and thus Deployment) to do what it needs to do.
+
+There are two things we'd like to improve about the way this works today:
+
+1. Folks want more control over how their Providers are deployed - over the
+   configuration of the Deployment (etc) Crossplane creates. ([#3601])
+2. In some cases, folks don't want Crossplane to create a Kubernetes Deployment
+   at all. They want to run a Provider some other way. ([#2671])
+
+Soon [Composition Functions][functions-beta-design] will also be deployed by the
+package manager as Kubernetes Deployments. We expect they'll have similar
+requirements. Since Functions aren't Kubernetes controllers, in this document
+I'll refer to the long-running processes some packages need to deploy as
+'package runtimes'.
+
+Crossplane has a v1alpha1 ControllerConfig type that addresses the first issue
+for Providers. It has been marked deprecated, to be removed if and when we find
+a suitable replacement. We deprecated ControllerConfig because:
+
+* It was growing piecemeal to support templatizing an entire Deployment.
+* We think in some rare cases package runtimes won't use Deployments.
+
+It's worth noting that the desire to deploy a package runtime as anything other
+than a Kubernetes Deployment in the same cluster where Crossplane is running is
+_quite rare_. To my knowledge only Upbound currently has this requirement.
+
+## Goals
+
+The goals of this design are to:
+
+* Give folks full control over package runtime Deployments.
+* Make it possible to run package runtimes as something other than a Deployment.
+
+## Proposal
+
+I propose a new flag to Crossplane: `--package-runtime`. This flag would have
```

Comment:
because this flag is passed to core crossplane, it would seem that the decision for all packages to use a particular runtime is being made at the time crossplane is deployed - not at the time each package is deployed. Is it possible to mix and match though, e.g. one package uses a deployment and the next package uses something external?

I **think** that may be possible reading further down because this new flag to crossplane more so sets the default runtime for packages that do not have a runtime set.  If that is correct (I'm not sure though, sorry 😇), would it make more sense to call this flag `--package-runtime-default`? 🤔

Reply:
I was thinking no - it would be a global setting affecting all packages.

Reply:
I was thinking about this as well, primarily in the context of having Crossplane manage functions "close" to itself, while deferring providers to an external system. However, there is a nice escape hatch in this proposal in that you always have an option to just say `--package-runtime=External` and implement whatever you want, which could include replicating the core Crossplane behavior for some packages and not for others.